sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/UploadBufferPool.java

```diff
@@ -0,0 +1,172 @@
+                // This will force us to get a new buffer next time we try to write.
+                this.currentBuf = null;
+            } else {
+                /*
+                We are still filling the current buffer, so we have no data to return. We will return the buffer once it
+                is filled
+                 */
+                result = Flux.empty();
+            }
+        } else {
+            // We will overflow the current buffer and require another one.
+            // Adjust the window of buf so that we fill up currentBuf without going out of bounds.
+            int oldLimit = buf.limit();
+            buf.limit(buf.position() + this.currentBuf.remaining());
+            this.currentBuf.put(buf);
+            // Set the old limit so we can read to the end in the next buffer.
+            buf.limit(oldLimit);
+
+            // Reset the position so we can read the buffer.
+            this.currentBuf.position(0);
+            result =  Flux.just(this.currentBuf);
+
+            /*
+            Get a new buffer and fill it with whatever is left from buf. Note that this relies on the assumption that
+            the source Flux has been split up into buffers that are no bigger than chunk size. This assumption
+            means we'll only have to over flow once, and the buffer we overflow into will not be filled. This is the
+            buffer we will write to on the next call to write().
+             */
+            this.currentBuf = this.getBuffer();
+            this.currentBuf.put(buf);
+        }
+        return result;
+    }
+
+    /*
+    Note that the upload method will be calling write sequentially as there is only one worker reading from the source
+    and calling write. Hence there is only one worker calling getBuffer at any time.
+     */
+    private ByteBuffer getBuffer() {
+        ByteBuffer result;
+        // There are no buffers in the queue and we have space to allocate one.
+        if (this.buffers.isEmpty() && this.numBuffs < this.maxBuffs) {
+            result = ByteBuffer.allocate(this.buffSize);
+            this.numBuffs++;
+        } else {
+            try {
+                // If empty, this will wait for an upload to finish and return a buffer.
+                result = this.buffers.take();
+
+            } catch (InterruptedException e) {
+                throw logger.logExceptionAsError(new IllegalStateException("BufferedUpload thread interrupted." + " Thread:"
+                        + Thread.currentThread().getId()));
+            }
+        }
+        return result;
+    }
+
+    Flux<ByteBuffer> flush() {
+        /*
+        Prep and return any data left in the pool. It is important to set the limit so that we don't read beyond the
+        actual data as this buffer may have been used before and therefore may have some garbage at the end.
+         */
+        if (this.currentBuf != null) {
+            this.currentBuf.flip();
```

Comment:
I think we want `flip`. If the customer only wrote 10 bytes at the end, we want to make sure we only read those 10 bytes rather than leaving the limit at the end of the buffer and sending extra data.